images/kindnetd/cmd/kindnetd/routes.go

```diff
@@ -24,30 +24,49 @@ import (
 	"k8s.io/klog/v2"
 )
 
-func syncRoute(nodeIP string, podCIDRs []string) error {
+func syncRoutes(nodeIP string, podCIDRs []string) error {
 	ip := net.ParseIP(nodeIP)
-
 	for _, podCIDR := range podCIDRs {
-		// parse subnet
-		dst, err := netlink.ParseIPNet(podCIDR)
-		if err != nil {
+		if err := syncRoute(podCIDR, ip); err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		// Check if the route exists to the other node's PodCIDR
-		routeToDst := netlink.Route{Dst: dst, Gw: ip}
-		route, err := netlink.RouteListFiltered(nl.GetIPFamily(ip), &routeToDst, netlink.RT_FILTER_DST)
-		if err != nil {
-			return err
+func syncRoute(podCIDR string, gw net.IP) error {
+	// parse subnet
+	dst, err := netlink.ParseIPNet(podCIDR)
+	if err != nil {
+		return err
+	}
+
+	// List all routes to the other node's PodCIDR
+	routeToDst := netlink.Route{Dst: dst, Gw: gw}
+	routes, err := netlink.RouteListFiltered(nl.GetIPFamily(gw), &routeToDst, netlink.RT_FILTER_DST)
+	if err != nil {
+		return err
+	}
+
+	// Check if the route exists to the other node's PodCIDR and delete invalid routes
+	for _, route := range routes {
+		// Skip creation of a route if there is gateway which matches the expected.
```

Comment:
I think we should just continue, i.e., let's say the route is 10.0.0.0/8 and the new gw 192.168.1.1, if we have
```
10.0.0.0/8 via 192.168.1.1
10.0.0.0/8 via 192.168.1.5
```
despite we have the right route we want to delete the other route too

Reply:
you're right 👍

Reply:
I think we can make the func easier.

We can expect to only have one route to our dst right?

So instead of using a for loop here, we could check for the length and have the following 4 cases:

* `len(routes) > 1`: --> error / unexpected
* `len(routes) == 1 && equal`: --> return nil
* `len(routes) == 1 && not equal`: --> remove and add route
* `len(routes) == 0`: add route

Reply:
Another solution:
 
 ave two calls to `netlink.RouteListFiltered`
 * the first to check if a route exists which matches gateway and dst: if so we can return early
 * the second to get the list of routes which only match dst: to remove them all
 * as third step add the new route

Reply:
the loop is not expensive and is easier to read, specially for people not used to networks. Also, we can solve the "routes > 1" , it is very cheap 

```
// myroute is the route we want to install
found := false
for _, r := routes {
   if r == myroute {
        found = true
   } else {
       delete(r)
   }
}
if !found {
  createRoute(myroute)
}
```

kindnet is a "toy" CNI, I prefer to keep it simple so "new" people to CNIs can

Reply:
Updated the parts to use continue and the boolean variable 👍